### ISSUE TITLE
DSLX: bug: dimension parametric expressions are very limited

### ISSUE THREAD

**Issue by rdob-ant** [google/xls#967]
Consider the following samples. Errors are specified in the comments inline.

**Sample 1 - using ArrayTypeAnnotation for fn return type**

```
fn myfunc1<NUMB: u32>() -> u8[NUMB] {
    u8[NUMB]:[u8:3, ...]
}
// ^^^ ok


fn myfunc2<NUMB: u32>() {
    let _ = u8[u32:4 * NUMB]:[u8:3, ...];
}
// ^^^ ok


fn myfunc3<NUMB: u32>() -> u8[u32:4 * NUMB] {
    u8[u32:4 * NUMB]:[u8:3, ...]
}
// ^^^
// Upon invokation from myfunc_concrete(), deducing return type annotation
// `-> u8[u32:4 * NUMB]` leads to following error:
//   TypeInferenceError: <> Could not evaluate dimension expression to a constant value:
//   INTERNAL: BytecodeEmitter could not find slot or binding for name: NUMB


fn myfunc_concrete() {
    let _ = myfunc1<u32:10>();
    let _ = myfunc2<u32:10>();
    let _ = myfunc3<u32:10>();
}
```

**Sample 2 - using ArrayTypeAnnotation for struct member**
```
struct Dummy {
    field: u32[4]
}

struct MyStruct2<WIDTH: u32> {
    a: bits[4],
    // ^^^ ok

    b: bits[WIDTH],
    // ^^^ ok

    c: Dummy[4],
    // ^^^ ok

    d: Dummy[WIDTH],
    // ^^^ ok

    e: bits[u32:4 * WIDTH],
    // ^^^ TypeInferenceError: <> Could not evaluate dimension expression to a constant value: INTERNAL: BytecodeEmitter could not find slot or binding for name: WIDTH

    f: bits[4 * WIDTH],
    // ^^^ TypeInferenceError: <> Could not infer a type for this number, please annotate a type.

    g: Dummy[u32:4 * WIDTH],
    // ^^^ TypeInferenceError: <> Could not evaluate dimension expression to a constant value: INTERNAL: BytecodeEmitter could not find slot or binding for name: WIDTH
}

// As one can see, same error happens both for arrays of built-in type (bits) as well as arrays of structs
```

My understanding is as follows.

In **Sample 1**, type-checking of invocation of parametric function `myfunc3` fails. During type-check of the invocation, values of all parametrics are expected to be defined, so all dimension expressions can be simply evaluated as constexprs. For some reason, though, the value of parametric `NUMB` is not visible to the _ConstexprEvaluator_ at the moment when _ArrayTypeAnnotation_ in the return type is deduced (I suppose the failure occurs when `deduce.cc:DimToConcrete()` attempts to evaluate dimension expression as constexpr).

In `myfunc2` the very same _ArrayTypeAnnotation_ is used in the function body, where it is deduced correctly.

`myfunc1` is also interesting - although it looks very similar to `myfunc3`, and the value of parametric is still undefined in that context and attempting to evaluate dimension as a constexpr would've surely failed, I think it's the `DimToConcrete()` who lets it slide as it has special handling for the case when dimension expression is a plain _NameRef_ - instead of calling _ConstexprEvaluator_, it just transforms such dimension expression into a *ParametricSymbol*, thus keeping it in symbolic form (which is probably resolved to an integer value at a later point).

In **Sample 2** we see a different aspect of the same mechanism: the errors happen during type-checking of parametric struct definitions. Unlike function invocations, values of parametrics are not known in this context, so dimension expressions can not be evaluated as constexprs - instead, they have to be transformed into ParametricExpressions and kept in symbolic form (similar to how `myfunc1` from *Sample 1* was treated). However, `DimToConcrete()`, which is responsible for this, does not implement elaborate expression conversion - if incoming dimension expression is neither a *Number* nor a *NameRef*, it just calls *ConstexprEvaluator* which fails as there is no way those parametrics can be resolved to concrete values at that point.

I've thought a bit about possible solutions.

For the **Sample 1**, it feels like it can be relatively easy to fix - we just need to propagate a known parametric value to a *ConstexprEvaluator*. Probably there is something wrong with scoping there, as it's already resolvable in the function body, but not when deducing return type annotation. 

For the **Sample 2**, perhaps writing proper code to convert dimension expression from _Expr AstNode_ to a _ParametricExpression_ will solve it?

And wrt the last thing, I have a more general question - why does DSLX interpreter attempt to type-check parametric structs at the point of definition in the first place? Why doesn't it postpone the checking till the point where those structs are instantiated or referenced from the context where all parametrics are defined, like it does for parametric functions? That would greatly simplify the flow, as all dimension expressions would be evaluatable as constexprs, potentially removing the need for _ParametricExpression_ altogether?

**Comment by cdleary** [comment]
The workaround is to used a "derived parametric" like the following, where you introduce a new parametric that's calculated based on other provided values -- that value is constexpr and a simple identifier:

```dslx
  fn f<NUMB: u32, D: u32 = {u32:4 * NUMB}>() -> u8[D] {
      u8[D]:[u8:3, ...]
  }
  fn main() -> u8[4] { f<u32:1>() }
```

Similar approach applies to parametric structs.

I'll work on doing better with these samples to create a more uniform experience (in general as a user you shouldn't be able to trigger internal errors so reports of those are always welcome) but that should generally unblock progress on creating examples using the frontend.
